### ISSUE TITLE
use FFHT code

### ISSUE THREAD

**Issue by stevengj** [stevengj/Hadamard.jl#5]
The https://github.com/FALCONN-LIB/FFHT implements a specialized SIMD FHT, which it says is faster than FFTW for this problem.  (Totally possible, since 2×2×2×⋯ multidimensional transforms are not that optimized in FFTW — they are using size-2 DFT kernels, whereas to be fast they should really use larger multidimensional kernels.)  Might be interesting to support using the FFHT (which is MIT-licensed) in cases where it is applicable.

**Comment by S6Regen** [comment]
When I ran the the FFHT bench-marking code on my cheap laptop I got 5200 65536-point WHT's per second per core.  In my code I get 5000 per core.

My legacy SSE code: https://drive.google.com/open?id=1bhQi4BybXPkvCENIYQcdw9PHAwS9Ocam

I see all the links I have given are broken if you just click on them.  It seems you have to copy and paste them to the browser address line for them to work!

Anyway to get GPU comparable results you would definitely need say one of the new AMD 32 core CPUs.  From the FFHT result tables that would get you to around 1,000,000 65536-point  WHT's per second.  A big advantage of CPUs is they are so much easier to program than GPUs.  I don't know what the situation is with the new GPU tensor cores.  If you could use them for the WHT you might reach 10 million+ 65536-point WHTs per second presuming the memory bandwidth could handle that.  That would be 1000 times better than I can do on my 2 core laptop (no harddrive, broken keyboard, CPU heatsink problems, runs off cheap usb stick etc.)  I could evolve a neural network in 86.4 seconds that takes me a day at present.

**Comment by Djoop** [comment]
I did a quick benchmark for the 1D case, it's probably not perfect but suggests that FFHT is one order of magnitude faster, even a bit more in low dimension. Here are the results (Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz).

![results](https://user-images.githubusercontent.com/6762779/48946332-a5db9c80-ef2d-11e8-81d2-eba6c741f999.png)

I create each time a matrix `A=randn(d,1000)`, and compare `@belapsed fwht_natural!($(A),dims=1)` and running `fht_double` on each column of A (with a simple loop, one can probably parallelize it.). They also provide an out-of-place version, which I guess is better than first copying.
For reference, there is also the Spiral version (http://www.spiral.net/software/wht.html), and this other SIMD C++ implementation : https://github.com/curto2/mckernel/ , but I didn't have time to compare.
So I don't know what is the best option : using FFHT only, using FFHT+FFTW for unsuported cases (maybe multidimensional cases?), or even creating a different package for FFHT?

**Comment by stevengj** [comment]
I would continue to use FFTW (because it probably handles more cases), and use FFHT for the cases it supports.

**Comment by stevengj** [comment]
Note that one problem with the above benchmarks is that currently it is timing the FFTW plan creation, which is slow — FFTW is designed by creating the plan first and then re-using it if you care about performance.  See #32.

Now that Julia supports easy external binary dependencies, however, it would be *much* easier to support calling FFHT than when this issue was first created.

**Comment by jofrevalles** [comment]
Hi! We were looking into a good implementation of the Fast Hadamard Transform (FHT) for our library [`HadaMAG.jl`](https://github.com/bsc-quantic/HadaMAG.jl), which mainly uses the FHT as a core function.
Nevertheless, we found that the implementation here in `Hadamard.jl` is quite slow, even slower than our simple pure Julia implementation. Maybe the problem is this plan creation timing issue you mention, but I have not found an instance in which this implementation of FHT worked better.
Therefore, we are using [`FastHadamardStructuredTransforms_jll`](https://github.com/JuliaBinaryWrappers/FastHadamardStructuredTransforms_jll.jl), which provides the binary wrappers to call the FFHT implementation. However, the problem is that this `FastHadamardStructuredTransforms_jll` does not package the ARM implementation of FFHT. If you could provide with some of these FFHT functions for a more general set of platforms, we would be very interested on using `Hadamard.jl`.

Thanks!
Jofre

**Comment by stevengj** [comment]
Yeah, I really need to fix #32 … up to now it hasn't been a priority since not many people seemed to be using this package.
